src/fixate/core/switching.py

```diff
@@ -0,0 +1,743 @@
+            tree, generate_bit_sets(pins_at_this_level)
+        ):
+            if signal_or_tree is None:
+                continue
+            if isinstance(signal_or_tree, Signal):
+                signal_map[signal_or_tree] = frozenset(pins_for_signal) | fixed_pins
+            else:
+                signal_map.update(
+                    self._map_tree(
+                        tree=signal_or_tree,
+                        pins=pins[bits_at_this_level:],
+                        fixed_pins=frozenset(pins_for_signal) | fixed_pins,
+                    )
+                )
+
+        return signal_map
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}('{self._state}')"
+
+    @staticmethod
+    def _default_update_pins(
+        pin_updates: PinUpdate, trigger_update: bool = True
+    ) -> None:
+        """
+        Output callback to effect a state change in the mux.
+
+        This is a default implementation which simply prints the planned state change to.
+        stdout. When instantiated as part of a jig driver, this will end up connected
+        to an AddressHandler to do the actual pin changes in hardware.
+
+        In general, this method shouldn't be overridden in a subclass. An alternative
+        can be provided to __init__.
+        """
+        print(pin_updates, trigger_update)
+
+
+class VirtualSwitch(VirtualMux):
+    """
+    A VirtualMux that controls a single pin.
+
+    A virtual switch is a multiplexer with a single pin. The multiplex
+    function can accept either boolean values or the strings 'TRUE' and
+    'FALSE'. The virtual address used to switch can be defined as a list
+    with the single element (as if it were a multiplexer) or by using the
+    shorthand which is to define the pin_name attribute as a string.
+    """
+
+    pin_name: Pin = ""
+    map_tree = ("FALSE", "TRUE")
+
+    def multiplex(
+        self, signal_output: Union[Signal, bool], trigger_update: bool = True
+    ) -> None:
+        if signal_output is True:
+            signal = "TRUE"
+        elif signal_output is False:
+            signal = "FALSE"
+        else:
+            signal = signal_output
+        super().multiplex(signal, trigger_update=trigger_update)
+
+    def __init__(
+        self,
+        update_pins: Optional[PinUpdateCallback] = None,
+    ):
+        if not self.pin_list:
+            self.pin_list = [self.pin_name]
+        super().__init__(update_pins)
+
+
+class RelayMatrixMux(VirtualMux):
+    clearing_time = 0.01
+
+    def _calculate_pins(
+        self, old_signal: Signal, new_signal: Signal
+    ) -> tuple[PinSetState, PinSetState]:
+        """
+        Override of _calculate_pins to implement break-before-make switching.
+        """
+        setup = PinSetState(off=self._pin_set, on=frozenset())
+        on_pins = self._signal_map[new_signal]
+        final = PinSetState(off=self._pin_set - on_pins, on=on_pins)
+        return setup, final
+
+
+class AddressHandler:
```

Comment:
Make this a Protocol if it only exists to document the required methods of a handler?

Reply:
Should all these be moved to handlers.py?

Reply:
> Make this a Protocol if it only exists to document the required methods of a handler?
Yeah, that probably makes sense.

Might be worth looking all the classes once we're happy with the general concept. In my examples at the moment, I've got sub-classing, sub-classing with `@dataclass`, potentially throw protocols in the mix. Hopefully doesn't end up too confusing with all the slight differences...

> Should all these be moved to handlers.py?
I don't think so, but agree it's a bit non-obvious what is going on. The handlers in switching.py don't do any real IO. they are  abstract intended to assist with making useful "real world" handers. The `RelayMatrixHandler` relies on drivers, which is why I moved it there. I think it makes sense for it to live there.

One specific issue I had was that you can't import drivers.ftdi without the driver. After moving `RelayMatrixHandler` in then occurred to me that is might make sense to have concrete vs abstract handers in different place. I'm not too fixed on that idea though

Reply:
`AddressHandler` has moved on a little since this and makes a bit more sense to keep as a subclass rather than switch to a protocol. Also, I suspect we will very rarely need to actually create new sub classes since we will almost always use `fixate.drivers.handlers.FTDIAddressHandler` now.